src/lib.rs

```diff
@@ -303,3 +191,167 @@ fn generate_analysis(config: &Config) -> Result<(), Box<std::error::Error>> {
 
     Ok(())
 }
+
+#[derive(Debug)]
+struct DocData {
+    krate: Crate,
+    data: HashMap<String, Item>,
+}
+
+impl DocData {
+    fn new(config: &Config) -> Result<DocData, Box<std::error::Error>> {
+        let roots = config.host.def_roots()?;
+
+        // FIXME: this whole code shouldn't look for a precise crate.
+        let root_id = roots.iter().find(|&&(_, ref name)| name == "example");
+        let root_id = match root_id {
+            Some(&(id, _)) => id,
+            _ => return Err(Box::new(CrateErr::new("example"))),
+        };
+
+        let root_def = config.host.get_def(root_id)?;
+
+        let name_len = root_def.qualname.len();
+        let mut krate = Crate {
+            id: root_id,
+            // example:: -> example
+            name: root_def.qualname[..(name_len - 2)].to_string(),
+            docs: root_def.docs.clone(),
+            modules: Vec::new(),
+        };
+
+        let data = DocData::build_data(config, root_id, &mut krate)?;
+
+        Ok(DocData { krate, data })
+    }
+
+    fn build_data(
+        config: &Config,
+        root_id: analysis::Id,
+        krate: &mut Crate,
```

Comment:
I think the reason this feels awkward is because passing in a mutable struct like this seems to emphasize some greater role in the `build_data` function when its current use is minimal. Looking at the resulting `data` HashMap, I think it might be interesting to explore encoding this data directly into the `Crate` struct. I'm not sure if that would be pushing too much information onto the `Crate` though. I don't think that change would need to be in this PR but just throwing out some food for thought.